automation/tests-container-utils.sh

```diff
@@ -0,0 +1,126 @@
+}
+
+function open_shell {
+    res=$?
+    [ "$res" -ne 0 ] && echo "*** ERROR: $res"
+    set +o errexit
+    container_exec 'echo "pytest tests/integration --pdb" >> ~/.bash_history'
+    container_exec 'exec /bin/bash'
+    run_exit
+}
+
+# Return 0 if file is changed else 1
+function is_file_changed {
+    git remote add upstream https://github.com/nmstate/nmstate.git
+    git fetch upstream
+    if [ -n "$TRAVIS_BRANCH" ]; then
+        git diff --exit-code --name-only upstream/$TRAVIS_BRANCH -- $1
+    else
+        git diff -exit-code --name-only upstream/master -- $1
+    fi
+
+    if [ $? -eq 0 ];then
+        return 1
+    else
+        return 0
+    fi
+}
+
+function rebuild_container_images {
+    set +x
+    if is_file_changed "$PROJECT_PATH/packaging"; then
+        if [ $CONTAINER_IMAGE == $CENTOS_IMAGE_DEV ];then
+            ${PROJECT_PATH}/packaging/build-container.sh centos8-nmstate-dev
+        elif [ $CONTAINER_IMAGE == $FEDORA_IMAGE_DEV ];then
+            ${PROJECT_PATH}/packaging/build-container.sh fedora-nmstate-dev
+        fi
+    fi
+}
+
+function remove_tempdir {
+    rm -rf "$NMSTATE_TEMPDIR"
+}
+
+function vdsm_tests {
+    trap remove_tempdir EXIT
+    git -C "$NMSTATE_TEMPDIR" clone --depth 1 https://gerrit.ovirt.org/vdsm
+    cd "$NMSTATE_TEMPDIR/vdsm"
+    ./tests/network/functional/run-tests.sh --nmstate-source="$PROJECT_PATH"
+    cd -
+}
+
+function collect_artifacts {
+    container_exec "
+      journalctl > "$CONT_EXPORT_DIR/journal.log" && \
+      dmesg > "$CONT_EXPORT_DIR/dmesg.log" && \
+      cat /var/log/openvswitch/ovsdb-server.log > "$CONT_EXPORT_DIR/ovsdb-server.log" && \
+      mv "$CONTAINER_WORKSPACE/pytest-run.log" "$CONT_EXPORT_DIR/pytest-run.log" || true
+    "
+}
+
+function prepare_container_environment {
+    ${CONTAINER_CMD} --version && cat /etc/resolv.conf
+
+    if [[ "$CI" == "true" ]];then
+        rebuild_container_images
+    fi
+
+    mkdir -p $EXPORT_DIR
+    CONTAINER_ID="$(${CONTAINER_CMD} run --privileged -d -e CI=$CI -v /sys/fs/cgroup:/sys/fs/cgroup:ro -v $PROJECT_PATH:$CONTAINER_WORKSPACE -v $EXPORT_DIR:$CONT_EXPORT_DIR $CONTAINER_IMAGE)"
+    [ -n "$debug_exit_shell" ] && trap open_shell EXIT || trap run_exit EXIT
+
+    if [[ -v customize_cmd ]];then
+        container_exec "${customize_cmd}"
+    fi
+
+    container_exec "echo '$CONT_EXPORT_DIR/core.%h.%e.%t' > \
+        /proc/sys/kernel/core_pattern"
+    container_exec "ulimit -c unlimited"
+}
+
+function copy_workspace_container {
+    if [[ "$CI" != "true" ]];then
+        if $CONTAINER_CMD --version | grep -qv podman; then
+          container_exec "cp -rf $CONTAINER_WORKSPACE /root/nmstate-workspace || true"
+        else
+          container_exec "cp -rf $CONTAINER_WORKSPACE /root/nmstate-workspace"
+        fi
+        # Change workspace to keep the original one clean
+        CONTAINER_WORKSPACE="/root/nmstate-workspace"
+    fi
+}
+
+function pass_env_var {
```

Comment:
This is not passing environment but actually start the container.

And you have two `docker run` in this script which is incorrect.

Reply:
I agree about it is starting the container. But sorry, I am not able to find the `docker run` that you mentioned.

Reply:
I've renamed the function to `create_container`. Thanks!